pkg/controller/apiextensions/requirement/reconciler.go

```diff
@@ -0,0 +1,453 @@
+	// that has no resource ref or class ref, so we can't assume the class ref
+	// is always set at this point.
+	if !meta.WasCreated(composite) {
+
+		if err := r.composite.Configure(ctx, requirement, composite); err != nil {
+			// If we didn't hit this error last time we'll be requeued
+			// implicitly due to the status update. Otherwise we want to retry
+			// after a brief wait, in case this was a transient error or some
+			// issue with the resource class was resolved.
+			log.Debug("Cannot configure composite resource", "error", err, "requeue-after", time.Now().Add(aShortWait))
+			record.Event(requirement, event.Warning(reasonCannotConfigureResource, err))
+			requirement.SetConditions(v1alpha1.Creating(), v1alpha1.ReconcileError(err))
+			return reconcile.Result{RequeueAfter: aShortWait}, errors.Wrap(r.client.Status().Update(ctx, requirement), errUpdateRequirementStatus)
+		}
+
+		// We'll know our composite resource's name at this point because it was
+		// set by the above configure step.
+		record = record.WithAnnotations("composite-name", composite.GetName())
+		log = log.WithValues("composite-name", composite.GetName())
+
+		if err := r.composite.Create(ctx, requirement, composite); err != nil {
+			// If we didn't hit this error last time we'll be requeued
+			// implicitly due to the status update. Otherwise we want to retry
+			// after a brief wait, in case this was a transient error.
+			log.Debug("Cannot create composite resource", "error", err, "requeue-after", time.Now().Add(aShortWait))
+			record.Event(requirement, event.Warning(reasonCannotCreateResource, err))
+			requirement.SetConditions(v1alpha1.Creating(), v1alpha1.ReconcileError(err))
+			return reconcile.Result{RequeueAfter: aShortWait}, errors.Wrap(r.client.Status().Update(ctx, requirement), errUpdateRequirementStatus)
+		}
+
+		log.Debug("Successfully created composite resource")
+		record.Event(requirement, event.Normal(reasonCreatedResource, "Successfully created composite resource"))
+	}
+
+	// TODO(negz): Check for TypeReady, not TypeSynced. Composite resources do
+	// not yet set the latter, but their connection secret is probably not very
+	// useful until the resources they compose are ready.
+	if !resource.IsConditionTrue(composite.GetCondition(v1alpha1.TypeSynced)) {
+		log.Debug("Composite resource is not yet ready")
+		record.Event(requirement, event.Normal(reasonWaitingToBind, "Composite resource is not yet ready"))
+
+		// TODO(negz): Watch the kind of composite resource we're concerned with
+		// so that we don't need to requeue here.
+		requirement.SetConditions(Waiting(), v1alpha1.ReconcileSuccess())
+		return reconcile.Result{RequeueAfter: aShortWait}, errors.Wrap(r.client.Status().Update(ctx, requirement), errUpdateRequirementStatus)
+	}
+
+	if composite.GetRequirementReference() == nil {
+		if err := r.composite.PropagateConnection(ctx, requirement, composite); err != nil {
```

Comment:
I think propagation should happen only after it successfully binds to the composite.